src/test/java/net/oauth/jsontoken/AsyncJsonTokenParserTest.java

```diff
@@ -141,6 +146,56 @@ public void testVerifyAndDeserialize_tokenFromRuby() throws Exception {
     assertEquals("world", token.getParamAsPrimitive("hello").getAsString());
   }
 
+  public void testDeserialize_corruptHeader() throws Exception {
+    AsyncJsonTokenParser parser = getAsyncJsonTokenParser();
+    InvalidJsonTokenException e = assertErrorCode(
+        ErrorCode.MALFORMED_TOKEN_STRING,
+        () -> parser.deserialize(TOKEN_STRING_CORRUPT_HEADER)
+    );
+
+    assertTrue(e.getCause() instanceof JsonParseException);
+  }
+
+  public void testVerifyWithVerifiers_unknownCause() throws Exception {
+    AsyncJsonTokenParser parser = getAsyncJsonTokenParser(asyncLocators, new AlwaysFailChecker());
+    JsonToken checkToken = naiveDeserialize(TOKEN_STRING);
+    assertErrorCode(
+        ErrorCode.UNKNOWN,
+        () -> parser.verify(checkToken, getVerifiers())
+    );
+  }
+
+  public void testVerifyWithVerifiers_nullSignature() throws Exception {
+    AsyncJsonTokenParser parser = getAsyncJsonTokenParser();
+    JsonToken checkToken = naiveDeserialize(TOKEN_STRING_2PARTS);
+    assertErrorCode(
+        ErrorCode.MALFORMED_TOKEN_STRING,
+        () -> parser.verify(checkToken, getVerifiers())
+    );
+  }
+
+  public void testSignatureIsValid_nullSignature() throws Exception {
+    AsyncJsonTokenParser parser = getAsyncJsonTokenParser();
+    assertErrorCode(
+        ErrorCode.MALFORMED_TOKEN_STRING,
+        () -> parser.signatureIsValid(TOKEN_STRING_2PARTS, getVerifiers())
+    );
+  }
+
+  /**
+   * Ensure that legitimate runtime exceptions get through rethrowing exceptions.
+   */
+  public void testRethrownException_runtimeException() throws Exception {
+    AsyncJsonTokenParser parser = getAsyncJsonTokenParser();
+    JsonToken checkToken = new JsonToken(new JsonObject());
+    ExecutionException e = assertThrows(
+        ExecutionException.class,
+        () -> parser.verify(checkToken).get()
+    );
+
+    assertTrue(e.getCause() instanceof IllegalStateException);
```

Comment:
nit: Can you also assert the message string of this IllegalStateException?

Reply:
Added